[MM-14534] Re-organize System Console into new sections

bradjcoughlin opened the pull request:
#### Summary
* Reorganizes system console settings into a flatter structure (2-tiered as opposed to 3).
* Hides page links in LHS if 'RestrictedSystemAdmin' setting is enabled in config.
* Routing redirects to default page if page is hidden in LHS
* Default page logic determines if license page is hidden and will redirect to next available page.

#### Ticket Links
https://mattermost.atlassian.net/browse/MM-14534
https://mattermost.atlassian.net/browse/MM-14535
https://mattermost.atlassian.net/browse/MM-14536

#### Related Pull Requests
N/A

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spintmint test server creation failed. See the logs for more information.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-068139c5f73a2f64a.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.3.59
Instance ID: i-068139c5f73a2f64a

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-03a31333b3e9fb944.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.36.238
Instance ID: i-03a31333b3e9fb944

bradjcoughlin commented:
@wiersgallak @jasonblais spinmint is up if you'd like to review the system console updates!

jasonblais commented on i18n/en.json:
```diff
@@ -482,6 +467,7 @@
   "admin.experimental.experimentalEnableDefaultChannelLeaveJoinMessages.title": "Enable Default Channel Leave/Join System Messages:",
   "admin.experimental.experimentalEnableHardenedMode.desc": "Enables a hardened mode for Mattermost that makes user experience trade-offs in the interest of security. See [documentation](!https://docs.mattermost.com/administration/config-settings.html#enable-hardened-mode-experimental) to learn more.",
   "admin.experimental.experimentalEnableHardenedMode.title": "Enable Hardened Mode:",
+  "admin.experimental.experimentalFeatures": "",
```
Is this intentionally blank?

bradjcoughlin replied:
Nope, good catch. The i18n string extraction tool wouldn't work directly in this case so I've had to manually search/replace those empty strings : )

jasonblais commented on i18n/en.json:
```diff
@@ -676,8 +657,10 @@
+  "admin.integrations.corsSettings": "CORS Settings",
+  "admin.integrations.gif": "GIF (Beta)",
+  "admin.integrations.integrationsFeatures": "Integrations Features",
+  "admin.integrations.plugins": "",
```
Is this intentionally blank?

bradjcoughlin commented:
@jasonblais 
1. Changed to "Plugins (Beta)"
2. "Plugins (Beta)" as the page title makes sense, changed
3. I thought "CORS" looked too short so I added settings 0/5
4. I'll see if I can move "Plugins" below "Gif" and "CORS".

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0dcb631cca2b1f6d3.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.36.37
Instance ID: i-0dcb631cca2b1f6d3

wiersgallak commented:
Thanks @bradjcoughlin - this is looking great.  Here are my notes: 

1. Category headings - "Integrations" and "Compliance" should be in all caps.  "Compliance Settings" should be "Compliance"
2. Logs:  Page heading is "Server Logs", Let's update the LHS section title to "Server Logs"
3. Database:  Page heading is "Database Settings", remove "Settings"
4. Image Proxy: Page heading is "ImageProxy", should be "Image Proxy (missing space)
5. Email Notifications: Page heading is "Email", add "Notifications"
6. Push Notifications: Page heading is "Mobile Push", Change to "Push Notifications"
7. Rate Limiting: Page heading is "Rate Limiting Settings", remove "Settings" 
8. Sign Up: Page heading is "Signup", change LHS section title to "Signup"
9. Email: Page heading is "Email Authentication", remove "Authentication"
10. Plugins: Page heading is "Management", update to "Plugins"
11. CORS: Page heading is "CORS Setting", remove "Settings"
12. Compliance Export - Update to LHS section title to "Compliance Export (Beta)
13. Compliance Monitoring: Page heading is "User Activity Logs" - anyway we can try to combine this with the settings on "Compliance Settings" page as first priority? Heading should be "Compliance Monitoring"

jespino commented on components/admin_console/admin_sidebar/admin_sidebar.jsx:
```diff
@@ -147,16 +147,29 @@ export default class AdminSidebar extends React.Component {
-                if (isVisible(item)) {
-                    result.add(section.url + '/' + item.url);
-                }
+        for (const item of Object.values(AdminDefinition.environment)) {
```
I think we can solve it with a loop inside another loop, right?

bradjcoughlin replied:
Yes, I'm actually refactoring a bit right now. Was waiting until content seemed solid.

bradjcoughlin replied:
So, maybe wait to review after a few hours : )

jespino commented on components/admin_console/admin_sidebar/admin_sidebar.jsx:
```diff
@@ -369,22 +390,15 @@ export default class AdminSidebar extends React.Component {
-                                    />
-                                }
-                            >
-                                {Object.values(AdminDefinition.settings).map(this.renderSettingsMenu)}
```
You can entirely remove the renderSettingsMenu funcion definition because was used only here.

bradjcoughlin replied:
Yep, I'm doing that now actually. Sorry, I tagged you a little prematurely. But good to confirm.

jespino commented on components/admin_console/admin_sidebar/admin_sidebar.jsx:
```diff
@@ -369,22 +390,15 @@ export default class AdminSidebar extends React.Component {
-                                {Object.values(AdminDefinition.settings).map(this.renderSettingsMenu)}
-                            </AdminSidebarCategory>
-                            {this.renderRootMenu(AdminDefinition.other, 'fa-wrench', 'admin.sidebar.other', 'OTHER')}
+                            {this.renderRootMenu(AdminDefinition.user_management, 'fa-users', 'admin.sidebar.userManagement', 'USER MANAGEMENT')}
```
As part of the renderRootMenu we take a look into all the items and check different things to only show valid entries. One check that was in the renderSettings menu and is not in renderRootMenu was:
```js
 if (!item.schema) {
    return;
}
```
Probably is a good thing to have there.

jespino commented on utils/admin_console_index.jsx:
```diff
@@ -44,19 +44,32 @@ function extractTextsFromSection(section, intl) {
 
 export function adminDefinitionsToUrlsAndTexts(adminDefinition, intl) {
     const entries = {};
+    for (const item of Object.values(adminDefinition.about)) {
```
Probably this can be solve entirely with a for inside a for.

bradjcoughlin replied:
I agree. I was thinking it could also be a small refactor ticket later.

jespino commented on components/admin_console/admin_console.jsx:
```diff
@@ -73,10 +73,20 @@ export default class AdminConsole extends React.Component {
         return (
             <Switch>
-                {Object.values({...AdminDefinition.reporting, ...AdminDefinition.other}).map((item) => {
+                {Object.values({
```
You here are removing the whole nesting thing so you are redirecting any invalid url to the first "about" section entry. The previous implementation using the renderSectionRoutes was handling that. Another possible solution to this would be something like:
```js
{Object.values(AdminDefinition).map(this.renderSectionRoutes.bind(this, extraProps))}
```
That inside the switch + the Redirect at the end, should work properly and as expected (I haven't tried)

bradjcoughlin replied:
I see that I still have a routing issue.

mattermod commented:
Spinmint test server destroyed

bradjcoughlin commented:
@wiersgallak we can discuss #12 further offline, but that's probably why I temporarily titled the "Compliance" page as "Compliance Settings" to differentiate.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0bbc36e2f8949ed39.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.48.13
Instance ID: i-0bbc36e2f8949ed39

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0d36fe8321d754dac.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.24.1
Instance ID: i-0d36fe8321d754dac

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-02319ea255751cc6b.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.0.125
Instance ID: i-02319ea255751cc6b

jasonblais commented:
@bradjcoughlin Sorry that I'm nitpicking but the title page for `GIF` is missing `(Beta)`, which was added to the sidebar 
![image](https://user-images.githubusercontent.com/13119842/55600003-489bf700-5728-11e9-8e9f-b8c203636820.png)

bradjcoughlin commented:
Not at all! Sorry I missed that.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0757bf95b9764d4b3.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.17.30
Instance ID: i-0757bf95b9764d4b3

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0926d5a4f02f0257f.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Internal IP: 172.26.62.6
Instance ID: i-0926d5a4f02f0257f

jespino commented on i18n/en.json:
```diff
@@ -162,28 +162,17 @@
   "add_user_to_channel_modal.title": "Add {name} to a channel",
   "add_users_to_team.title": "Add New Members To {teamName} Team",
   "admin.advance.cluster": "High Availability",
-  "admin.advance.experimental": "Experimental",
```
In this files, looks like there are 31 strings that has been removed from the original admin console. That means that we have less configuration than before. Maybe some of them is because the reordering, but 31 means that probably there are sections entirely removed.

bradjcoughlin replied:
I also find this unsettling, but I can't point to the exact reasons for the reduction in strings until unless I audit the entire file. I was manually comparing throughout development, but I believe your string extraction utility may have identified some outdated strings - I could be wrong. I will look this over manually once more.

bradjcoughlin replied:
Any place where we've combined 2 pages into a single page would remove the page title strings as well.

jespino replied:
Ok, then this difference makes sense.

jasonblais commented on components/admin_console/admin_console.jsx:
```diff
@@ -73,63 +74,48 @@ export default class AdminConsole extends React.Component {
-            />
+            <Switch>
+                {schemas}
+                {<Redirect to={`${this.props.match.url}/${defaultUrl}`}/>}
```
@bradjcoughlin would you know if the new behaviour is expected?

Also noticed the URL for some pages is `../admin_console/<feature>` instead of `../admin_console/<section>/<feature>`

bradjcoughlin replied:
@jespino I've noticed this as well. I'm going to create a subsequent ticket to address some of the routing and URL concerns.

bradjcoughlin replied:
@jasonblais @wiersgallak here is the follow-up ticket regarding URLs and routing: https://mattermost.atlassian.net/browse/MM-15027. I've noted that a decision needs to be made in regards to what should happen if a user browses to a URL that is hidden (in Environment for example), when the entire section is hidden. Should they be redirected to the first page in the system console?

jasonblais commented on i18n/en.json:
```diff
@@ -162,28 +162,17 @@
   "admin.authentication.saml": "SAML 2.0",
+  "admin.authentication.signup": "Signup",
   "admin.banner.heading": "Note:",
-  "admin.client_versions.androidLatestVersion": "Latest Android Version",
```
Yes (the page was never actually displayed in the admin console)

bradjcoughlin replied:
This accounts for 12 of the 31 removed lines. 19 remaining.

mattermod commented:
Spinmint test server destroyed
